### PR TITLE
API Guidelines

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR (Pull Request) do?
- [x] This PR is not related to any Issue

Adds the API guidelines previously merged in [a blueprint](https://github.com/apache/trafficcontrol/blob/master/blueprints/api.guidelines.md) to the official docs.

## Which Traffic Control components are affected by this PR?
- Documentation

## What is the best way to verify this PR?
No tests for docs; just build and read them.

## The following criteria are ALL met by this PR
- [x] I have explained why tests are unnecessary
- [x] This PR includes documentation
- [x] An update to CHANGELOG.md is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

### CODE REVIEW

**Comment by mitchell852** [comment]
@ocket8888 - want to link this to the contributing.md file?

File: docs/source/development/api_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+- ``warning`` - This level is used to warn clients of potentially dangerous conditions when said conditions have not caused a request to fail. The best example of this is deprecation warnings, which should appear on all API routes that have been deprecated.
+
+Summary
+-------
+The ``summary`` object is used to provide summary statistics about object collections. In general the use of ``summary`` is left to be defined by API endpoints (subject to some restrictions). However, its use is not appropriate in cases where the user is specifically requesting summary statistics, but should rather be used to provide supporting information - pre-calculated - about a set of objects or data that the client *has* requested.
+
+Endpoints MUST use the following, reserved properties of ``summary`` for their described purposes (when use of ``summary`` is appropriate) rather than defining new ``summary`` or ``response`` properties to suit the same purpose:
+
+- ``count`` - Count contains an unsigned integer that defines the total number of results that could possibly be returned given the non-pagination query parameters supplied by the client.
```
Comment (mitchell852) [inline docs/source/development/api_guidelines.rst:73]: did u want to mention min, max, mean here?

File: docs/source/development/api_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+When an object is replaced, the response body MUST contain a representation of the object after replacement. While :RFC:`2616` states that servers MAY create objects for the passed representations if they do not already exist, :ref:`to-api` endpoint authors MUST instead use POST handlers for object creation.
+
+All endpoints that support the PUT request method MUST also support the :mailheader:`If-Unmodified-Since` HTTP header.
```
Comment (mitchell852) [inline docs/source/development/api_guidelines.rst:97]: different than `If-Not-Modified-Since`?

File: docs/source/development/api_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+409 Conflict
+------------
+This SHOULD be used when the request cannot be completed because the current state of the server is fundamentally incompatible with the request. For example, creating a new user with an email that is already in use should result in this response.
+
+Additionally, this MAY be used instead of `404 Not Found`_ when the client is requesting a link between an object identified by the request URI and some other object (e.g. when assigning a :term:`cache server` to a :term:`Delivery Service`) when the other object does not exist. If the request URI identifies an object that does not exist, the response MUST use `404 Not Found`_ instead.
+
+This is also the proper response status code when the conditions of a request cannot be met, e.g. when a client submits a PATCH request for a resource with an :mailheader:`If-Match` header that does not match the stored object's :mailheader:`ETag`.
+
+The request body MUST indicate what the conflict is that prevented the request from being fulfilled via one or more ``error``-level alerts.
```
Comment (mitchell852) [inline docs/source/development/api_guidelines.rst:172]: response body?

File: docs/source/development/api_guidelines.rst (hunk 1 above)
Comment (ocket8888) [reply docs/source/development/api_guidelines.rst:73]: No, that's in the "Response" object section

File: docs/source/development/api_guidelines.rst (hunk 2 above)
Comment (ocket8888) [reply docs/source/development/api_guidelines.rst:97]: Yes: [one of them is real, and the other isn't](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers#Conditionals)

I get that wrong all the time too. In fact, it was wrong in the original blueprint until Rob pointed it out.

Comment (mitchell852) [reply docs/source/development/api_guidelines.rst:97]: so you're going to fix the `If-Not-Modified-Since` reference that i saw somewhere?

Comment (ocket8888) [reply docs/source/development/api_guidelines.rst:97]: Yeah, I'll need to. Where is it?

Comment (ocket8888) [reply docs/source/development/api_guidelines.rst:97]: nvm, I found it

**Pull request merged**